File: p872e/leaf_similar.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
// Generator version is much slower
var leafSimilar = function (root1, root2) {
  function* seqGenerator(root) {
    if (root == null) return;
    if (root.left == null && root.right == null) yield root.val;
    yield* seqGenerator(root.left);
    yield* seqGenerator(root.right);
  }
  let seq1 = seqGenerator(root1);
  let seq2 = seqGenerator(root2);
  let v1, v2;
  do {
    [v1, v2] = [seq1.next(), seq2.next()];
    if (v1.value !== v2.value) return false;
  } while (!v1.done && !v2.done);
  return v1.done && v2.done;
};
// TESTS
[
  {
    tree1: '3,#,#',
    tree2: '1,#,#',
    expected: false,
  },
  {
    tree1: '3,5,6,#,#,2,7,#,#,4,#,#,1,9,#,#,8,#,#',
    tree2: '3,5,6,#,#,7,#,#,1,4,#,#,2,9,#,#,8,#,#',
    expected: true,
  },
].forEach(({ tree1, tree2, expected }) => {
  const root1 = Tree.deserialize(tree1);
  const root2 = Tree.deserialize(tree2);
  const actual = leafSimilar(root1, root2);
  console.log('Two trees', tree1, tree2, 'are leaf similar ->', actual);
  console.assert(actual === expected);
});

File: p872e/leaf_similar.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from local_packages.binary_tree import TreeNode
from typing import Optional


class Solution:
    def leafSimilar(
        self, root1: Optional[TreeNode], root2: Optional[TreeNode]
    ) -> bool:
        def leafValueSeq(root: Optional[TreeNode]) -> list:
            if not root:
                return []
            if not root.left and not root.right:
                return [root.val]
            else:
                return leafValueSeq(root.left) + leafValueSeq(root.right)

        return leafValueSeq(root1) == leafValueSeq(root2)


# TESTS
for tree1, tree2, expected in [
    ("1,#,#", "1,#,#", True),
    ("1,2,#,#,#", "1,#,2,#,#", True),
    (
        "3,5,6,#,#,2,7,#,#,4,#,#,1,9,#,#,8,#,#",
        "3,5,6,#,#,7,#,#,1,4,#,#,2,9,#,#,8,#,#",
        True,
    ),
    ("1,2,#,#,3,#,#", "1,3,#,#,2,#,#", False),
]:
    r1, r2 = TreeNode.deserialize(tree1), TreeNode.deserialize(tree2)
    sol = Solution()
    actual = sol.leafSimilar(r1, r2)
    print("Tree1:", tree1, "and Tree2:", tree2, "are leaf similar ->", actual)
    assert actual == expected
